src/lib.rs

```diff
@@ -208,9 +208,108 @@
 // Ensure we're `no_std` when compiling for Wasm.
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use srml_support::traits::Currency;
-use srml_support::{decl_event, decl_module, decl_storage, dispatch, StorageValue};
-use system::ensure_signed;
+#[cfg(feature = "std")]
+#[macro_use]
+extern crate serde_derive;
+
+use rstd::prelude::*;
+
+use parity_codec_derive::{Decode, Encode};
+use srml_support::{ decl_event, decl_module, decl_storage, ensure, dispatch, StorageValue, StorageMap};
+
+/// Length constraint for input validation
+
+enum LengthValidationResult {
+    TooShort,
+    TooLong,
+    Success
+}
+
+struct InputValidationLengthConstraint {
+
+    /// Minimum length
+    min : usize,
+
+    /// Maximum length
+    /// While having max would have been more direct, this
+    /// way makes max < min unrepresentable semantically, 
+    /// which is safer.
+    max_as_min_diff: usize,
```

Comment:
Cannot say that I like this idea of diff. Looks like an overenginering to me. 
1. You afraid that somebody silently will change max so it will be smaller then min. But an attacker can otherwise change plus sign to minus in this code: `self.min + self.max_as_min_diff`
2. We never used such approach in any other module.
3. I haven't seen this approach in other software.
4. UI should replicate the same function for `self.min + self.max_as_min_diff`.
5. You can check that min < max in validation function if you want.

Reply:
I don't know if I will keep this `InputValidationLengthConstraint`, but I should respond to some of these points. Most importantly, I am not sure if you are just objecting to the idea of using the `max_as_min_diff` approach or more broadly having even this whole struct, even if `max` was explicitly represented. 

I will assume it is the first case.

1. I am not sure what you mean by "an attacker", which apparently can alter the source. Anyone who can alter the source can do anything, so it's a moot point.

The underlying point is to use the type system to make bad states unrepresentable, which is the entire point of type safety, and that is what this approach does. It makes it so that a developer cannot accidentally allow an invalid max/min combination, which could happen in many ways, depending on how these constraints are being set or changed (e.g. manually in code and fixed forever, or changed with calls from governance).

2. Ok.

3. This is because you are misunderstanding what the goal here is because this is done all the time. The simple choice of using an unsigned variable rather than a signed variable here is just the same basic point.

4. It will need to be able to read the values and interpret them correctly, yes. It may choose to explicitly store max/min in its own representation, but the whole question is what is actually a safer way to represent it. Whatever the answer is, it should probably be the same in both runtime and UI code, they should both be as safe/practical possible.

5. This is a runtime check, which is strictly less safe and more complicated to handle than having the type system statically make it impossible.

You can still have a bad state in your system that you somehow must recover from, and you will  have to have some sort of error handling in the code calling `validate`, otherwise, your system will still fail.

In conclusion, I am not sure if I keep this struct - but I think it should just stay for now, I feel like `LengthValidationResult` is the part that may be overengineered, but even if I drop both, I still think it's better to use the basic idea in the representation of the constraint itself.

Reply:
Here I am concerned about max_as_min_diff. The whole valuation structure and methods look fine.

Reply:
Ok, good to know.